[unwind.cpp] In process unwind symbolization

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108469

(WIP to debug build issues)
Replaces the Symbolizer which calls `addr2line` with one that uses
libbfd in process to do the same work. This is a little nicer to deal with
because it doesn't have to fork, which can be a problem for large applications.
It also doesn't keep subprocesses around that might not clean up nicely.

libbfd is the same library addr2line uses, so the results should be equivalent
and take about the same time to generate.

This uses `std::async` to launch threads because libbfd is pretty slow,
and doing so matches the performance of the previous version which launched
processes instead of threads.

Differential Revision: [D48980893](https://our.internmc.facebook.com/intern/diff/D48980893)

zdevito committed:
[unwind.cpp] In process unwind symbolization

(WIP to debug build issues)
Replaces the Symbolizer which calls `addr2line` with one that uses
libbfd in process to do the same work. This is a little nicer to deal with
because it doesn't have to fork, which can be a problem for large applications.
It also doesn't keep subprocesses around that might not clean up nicely.

libbfd is the same library addr2line uses, so the results should be equivalent
and take about the same time to generate.

This uses `std::async` to launch threads because libbfd is pretty slow,
and doing so matches the performance of the previous version which launched
processes instead of threads.

[ghstack-poisoned]

Skylion007 commented on torch/csrc/profiler/unwind/unwind.cpp:
```diff
@@ -398,6 +408,190 @@ struct Symbolizer {
+  }
+  const Frame& lookup(void* addr) {
+    if (has_pending_results_) {
+      std::vector<std::future<std::vector<Frame>>> futures;
```
```suggestion
      std::vector<std::future<std::vector<Frame>>> futures;
      futures.reserve(entries_.size());
```

Skylion007 commented on torch/csrc/profiler/unwind/unwind.cpp:
```diff
@@ -398,6 +408,190 @@ struct Symbolizer {
+        e.syms_.resize(real_size);
+      }
+    }
+    std::vector<Frame> results;
```
```suggestion
    std::vector<Frame> results;
    results.reserve(e.pending_.size());
```

Skylion007 reviewed (commented):
Preallocate vectors when possible

aaronenyeshi reviewed (changes requested):
Overall this looks great! Please take a look at  the CI build failures, probably missing the bfd library in some environments. Also, I wonder if this should be turned on by default or not.

zdevito commented:
The build errors for all the linux distros are going to be a problem. This is why the original code uses `addr2line`. It is almost always installed but the development files for libbfd which is uses are not. I'm going to see if I can get this to build internally and then only use this path there.

zdevito committed:
Update on "[unwind.cpp] In process unwind symbolization"


(WIP to debug build issues)
Replaces the Symbolizer which calls `addr2line` with one that uses
libbfd in process to do the same work. This is a little nicer to deal with
because it doesn't have to fork, which can be a problem for large applications.
It also doesn't keep subprocesses around that might not clean up nicely.

libbfd is the same library addr2line uses, so the results should be equivalent
and take about the same time to generate.

This uses `std::async` to launch threads because libbfd is pretty slow,
and doing so matches the performance of the previous version which launched
processes instead of threads.

[ghstack-poisoned]

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48980893).
